Decred Wallet/Features/Wallet Utils/SyncManager.swift

```diff
@@ -0,0 +1,152 @@
+//
+//  SyncManager.swift
+//  Decred Wallet
+//
+// Copyright (c) 2019 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+import Foundation
+import Dcrlibwallet
+import Signals
+
+
+class SyncManager: NSObject, SyncProgressListenerProtocol {
+    static let shared = SyncManager()
+    
+    var syncing = Signal<(Bool, String?)>()
+    let syncProgress =  Signal<(DcrlibwalletGeneralSyncProgress?, DcrlibwalletHeadersFetchProgressReport?)>()
+    var peers = Signal<Int32>()
+    var syncStage = Signal<(Int, String)>()
+    var connectedToNetwork: ((_ status: Bool) -> Void)?
+    
+    override init() {
+        super.init()
+        AppDelegate.walletLoader.syncer.registerSyncProgressListener(for: "\(self)", self)
+        AppDelegate.walletLoader.syncer.registerEstimatedSyncProgressListener()
+        self.connectedToNetwork?(false) // we assume wallet is offline on startup
+        self.checkNetworkConnectionForSync()
+    }
+    
+    func checkNetworkConnectionForSync() {
+        // Re-trigger app network change listener to ensure correct network status is determined.
+        AppDelegate.shared.listenForNetworkChanges()
+        
+        if AppDelegate.shared.reachability.connection == .none {
+            DispatchQueue.main.async {
+                let noConnectionAlert = UIAlertController(title: LocalizedStrings.internetConnectionRequired, message: LocalizedStrings.cannotSyncWithoutNetworkConnection, preferredStyle: .alert)
+                noConnectionAlert.addAction(UIAlertAction(title: LocalizedStrings.ok, style: .default, handler: nil))
+                AppDelegate.shared.window?.rootViewController?.present(noConnectionAlert, animated: false, completion: self.checkSyncPermission)
+                self.connectedToNetwork?(false)
+            }
+        } else {
+            self.checkSyncPermission()
+        }
+    }
+    
+    func checkSyncPermission() {
+        if AppDelegate.shared.reachability.connection == .none {
+            self.syncNotStartedDueToNetwork()
+        } else if AppDelegate.shared.reachability.connection == .wifi || Settings.syncOnCellular {
+            self.startSync()
+        } else {
+            self.requestPermissionToSync()
+        }
+    }
+    
+    func requestPermissionToSync() {
+        
```

Comment:
This empty line space is not really necessary. If we make it a practice to start all function bodies with empty line spaces, our code would become less than pretty quickly. Exceptions are for function signatures that are so long that they wrap into the next line - then an empty line space is necessary.

Reply:
resolved in aa27838

Reply:
resolved in aa27838